File: src/utils/pinyin.test.ts
// @vitest-environment node
import { describe, it, expect } from "vitest";
import { pinyinMatch } from "./pinyin";

describe("pinyinMatch", () => {
  it("empty query matches everything", () => {
    expect(pinyinMatch("", "anything")).toBe(true);
    expect(pinyinMatch("   ", "anything")).toBe(true);
  });

  it("direct substring match (case-insensitive)", () => {
    expect(pinyinMatch("GLM", "GLM-4")).toBe(true);
    expect(pinyinMatch("glm", "GLM-4")).toBe(true);
    expect(pinyinMatch("lian", "百炼")).toBe(true); // matches via target pinyin "bailian"
  });

  it("full pinyin of chinese target", () => {
    expect(pinyinMatch("bailian", "百炼")).toBe(true);
    expect(pinyinMatch("bai", "百炼")).toBe(true);
    expect(pinyinMatch("xiaomi", "小米")).toBe(true);
  });

  it("chinese query converted to pinyin then matched", () => {
    // target latin, query chinese → queryPinyin path
    expect(pinyinMatch("百", "百炼")).toBe(true);
    expect(pinyinMatch("炼", "百炼")).toBe(true);
  });

  it("mixed chinese + latin query", () => {
    expect(pinyinMatch("百lian", "百炼")).toBe(true);
    expect(pinyinMatch("xiao米", "小米")).toBe(true);
  });

  it("returns false when nothing matches", () => {
    expect(pinyinMatch("zzz", "百炼")).toBe(false);
  });

  it("non-chinese characters preserved during pinyin conversion", () => {
    expect(pinyinMatch("xiaomiai", "小米AI")).toBe(true);
  });

  it("LRU cache survives many distinct targets (eviction path)", () => {
    // 构造 > 500 个不同 target 触发淘汰，确认仍正确匹配（不崩、命中正常）
    for (let i = 0; i < 600; i++) {
      pinyinMatch("x", `平台${i}`);
    }
    // 早期 target 已被淘汰，重新查询应重算且正确
    expect(pinyinMatch("pingtai", "平台0")).toBe(true);
    // 缓存命中路径：再查一次同 target
    expect(pinyinMatch("pingtai", "平台0")).toBe(true);
  });
});

File: flutter/test/utils/pinyin_test.dart
/// `lib/src/utils/pinyin.dart` 的单测 —— 断言逐条翻译自 React 版
/// `src/utils/pinyin.test.ts`（pinyin-pro 语义，数据与期望值一字不改）。
library;

import 'package:aidog_flutter/src/utils/pinyin.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('pinyinMatch（pinyin.test.ts 逐条翻译）', () {
    test('空查询匹配一切', () {
      expect(pinyinMatch('', 'anything'), isTrue);
      expect(pinyinMatch('   ', 'anything'), isTrue);
    });

    test('直接子串（不分大小写）', () {
      expect(pinyinMatch('GLM', 'GLM-4'), isTrue);
      expect(pinyinMatch('glm', 'GLM-4'), isTrue);
      expect(pinyinMatch('lian', '百炼'), isTrue); // 命中全拼 bailian
    });

    test('中文目标转全拼', () {
      expect(pinyinMatch('bailian', '百炼'), isTrue);
      expect(pinyinMatch('bai', '百炼'), isTrue);
      expect(pinyinMatch('xiaomi', '小米'), isTrue);
    });

    test('中文 query 转拼音再匹配', () {
      expect(pinyinMatch('百', '百炼'), isTrue);
      expect(pinyinMatch('炼', '百炼'), isTrue);
    });

    test('中英混合', () {
      expect(pinyinMatch('百lian', '百炼'), isTrue);
      expect(pinyinMatch('xiao米', '小米'), isTrue);
    });

    test('都不命中 → false', () {
      expect(pinyinMatch('zzz', '百炼'), isFalse);
    });

    test('非中文字符原样保留', () {
      expect(pinyinMatch('xiaomiai', '小米AI'), isTrue);
    });
  });

  group('拼音首字母（pinyin.test.ts 逐条翻译）', () {
    test('bl → 百炼 / yzam → 月之暗面 / zp → 智谱 / xm → 小米', () {
      expect(pinyinMatch('bl', '百炼'), isTrue);
      expect(pinyinMatch('yzam', '月之暗面'), isTrue);
      expect(pinyinMatch('zp', '智谱'), isTrue);
      expect(pinyinMatch('xm', '小米'), isTrue);
    });

    test('首字母子串也命中（zh → 智谱）', () {
      expect(pinyinMatch('zh', '智谱'), isTrue);
    });

    test('拉丁 target 不走首字母分支', () {
      expect(pinyinMatch('zp', 'GLM'), isFalse);
      expect(pinyinMatch('gl', 'GLM'), isTrue);
    });

    test('含中文的 query 不触发首字母分支', () {
      expect(pinyinMatch('百', '智谱'), isFalse);
    });
  });

  group('词典边界', () {
    test('3500 常用字外：原样保留、不抛错', () {
      // 㵘（U+3D58）在《现代汉语常用字表》之外，词典查不到 → 按原字符处理。
      expect(pinyinMatch('㵘', '㵘测试'), isTrue);
      expect(toPinyin('㵘'), '㵘');
    });

    test('首字母串只含汉字的首字母，拉丁字符跳过', () {
      expect(toInitials('小米AI'), 'xm');
      expect(toPinyin('小米AI'), 'xiaomiai');
    });

    test('同一 target 重复转换走缓存，结果稳定', () {
      expect(pinyinMatch('fenzu', '分组'), isTrue);
      expect(pinyinMatch('fenzu', '分组'), isTrue);
      expect(pinyinMatch('fz', '分组'), isTrue);
    });
  });
}
